cpp/roaring.hh

```diff
@@ -18,38 +19,45 @@ class Roaring {
     /**
      * Create an empty bitmap
      */
-    Roaring() : roaring(NULL) {
-        roaring = roaring_bitmap_create();
-        if (roaring == NULL) {
+    Roaring() {
+        bool is_ok = ra_init(&roaring.high_low_container);
+        if (!is_ok) {
             throw std::runtime_error("failed memory alloc in constructor");
         }
+        roaring.copy_on_write = false;
     }
 
     /**
      * Construct a bitmap from a list of integer values.
      */
-    Roaring(size_t n, const uint32_t *data) {
-        roaring = roaring_bitmap_of_ptr(n, data);
-        if (roaring == NULL) {
-            throw std::runtime_error("failed memory alloc in constructor");
-        }
+    Roaring(size_t n, const uint32_t *data) : Roaring() {
+        roaring_bitmap_add_many(&roaring, n, data);
     }
     /**
      * Copy constructor
      */
-    Roaring(const Roaring &r) : roaring(NULL) {
-        roaring = roaring_bitmap_copy(r.roaring);
-        if (roaring == NULL) {
+    Roaring(const Roaring &r) {
+        bool is_ok = ra_copy(&r.roaring.high_low_container, &roaring.high_low_container,
+                             r.roaring.copy_on_write);
+        if (!is_ok) {
             throw std::runtime_error("failed memory alloc in constructor");
         }
+        roaring.copy_on_write = r.roaring.copy_on_write;
     }
 
     /**
      * Construct a roaring object from the C struct.
      *
      * Passing a NULL point is unsafe.
-     */
-    Roaring(roaring_bitmap_t *s) : roaring(s) {}
+     * the pointer to the C struct will be invalid after the call.
+     */
+    Roaring(roaring_bitmap_t *s) {
```

Comment:
This is an interesting case. It isn't possible to replicate exactly the old behavior of this function with an embedded roaring_bitmap_t because we have to use new memory in a constructor. I've selected the approach because it seems the least disruptive: There isn't a copy so it hasn't changed in terms of semantics or much in terms of performance. It frees the memory it is given so the contract hasn't changed, all of the memory is still now handled by the C++ object. The only downside is that the given pointer is no longer a valid alias to the data structure.